Kudu.SiteManagement/SiteManager.cs

```diff
@@ -374,92 +390,93 @@ private static int GetRandomPort(IIS.ServerManager iis)
-            }
-
-            return true;
+            return tcpConnections.All(connectionInfo => connectionInfo.LocalEndPoint.Port != port)
+                && iis.Sites
+                    .SelectMany(iisSite => iisSite.Bindings)
+                    .All(binding => binding.EndPoint == null || binding.EndPoint.Port != port);
         }
 
-        private static bool IsAvailable(string host, int port, IIS.ServerManager iis)
+        private static bool IsAvailable(string host, int port, ServerManager iis)
         {
-            foreach (var iisSite in iis.Sites)
-            {
-                foreach (var binding in iisSite.Bindings)
-                {
-                    if (binding.EndPoint != null && binding.EndPoint.Port == port && binding.Host == host)
-                    {
-                        return false;
-                    }
-                }
-            }
-
-            return true;
+            return iis.Sites
+                .SelectMany(iisSite => iisSite.Bindings)
+                .All(binding => binding.EndPoint == null || binding.EndPoint.Port != port || binding.Host != host);
         }
 
-        private IIS.Site CreateSiteAsync(IIS.ServerManager iis, string applicationName, string siteName, string siteRoot, List<string> siteBindings)
+        private IIS.Site CreateSiteAsync(ServerManager iis, string applicationName, string siteName, string siteRoot, List<SiteManager.BindingInformation> bindings)
         {
             var pool = EnsureAppPool(iis, applicationName);
 
             IIS.Site site;
-
-            if (siteBindings != null && siteBindings.Count > 0)
+            if (bindings.Any())
             {
-                site = iis.Sites.Add(siteName, "http", siteBindings.First(), siteRoot);
+                BindingInformation first = bindings.First();
+
+                //Binding primaryBinding;
+                site = first.Scheme == UriScheme.Http
+                    ? iis.Sites.Add(siteName, "http", first.Binding, siteRoot)
+                    : iis.Sites.Add(siteName, first.Binding, siteRoot, _certificateSearcher.Lookup(first.Certificate).ByFriendlyName().GetCertHash());
+                if (first.Configuration.RequireSni)
+                {
+                    site.Bindings.First().SetAttributeValue("sslFlags", SslFlags.Sni);
+                }
+
+                //Note: Add the rest of the bindings normally.
+                foreach (BindingInformation binding in bindings.Skip(1))
+                {
+                    switch (binding.Scheme)
+                    {
+                        case UriScheme.Http:
+                            site.Bindings.Add(binding.Binding, "http");
+                            break;
+
+                        case UriScheme.Https:
+                            Certificate cert = _certificateSearcher.Lookup(binding.Certificate).ByFriendlyName();
+                            Binding bind = site.Bindings.Add(binding.Binding, cert.GetCertHash(), cert.StoreName);
```

Comment:
Agreed, so for now I changed it to throw an ConfigurationErrorsException with a more meaning-full message, it feels a bit odd to throw that here, but it's the most meaning-full thing I could come up with at the moment.